lib/IRGen/GenDecl.cpp

```diff
@@ -6259,3 +6259,35 @@ void IRGenModule::setColocateMetadataSection(llvm::Function *f) {
     break;
   }
 }
+
+void IRGenModule::emitAsyncEntryReturnFunctionSection() {
+  if (TargetInfo.OutputObjectFormat != llvm::Triple::MachO ||
+      !getOptions().EmitAsyncFramePushPopMetadata)
+    return;
+
+  auto createSection = [&] (const char *globalName, const char *sectionName,
+                            SmallVectorImpl<llvm::Function*> &functions) {
+    if (functions.empty())
+      return;
+
+    ConstantInitBuilder builder(*this);
+    auto array = builder.beginArray(RelativeAddressTy);
+    for (auto *asyncFn : functions) {
+      array.addRelativeAddress(asyncFn);
+    }
+
+    auto var = array.finishAndCreateGlobal(
+      globalName , Alignment(4), /*isConstant*/ true,
+      llvm::GlobalValue::InternalLinkage);
+    var->setSection(sectionName);
+    disableAddressSanitizer(*this, var);
+    addUsedGlobal(var);
+  };
+
+  createSection("__swift_async_entry_functlets",
+                "__TEXT,__swift_as_entry, coalesced, no_dead_strip",
+                AsyncEntryFunctions);
+  createSection("__swift_async_ret_functlets",
+                "__TEXT,__swift_as_ret, coalesced, no_dead_strip",
+                AsyncReturnFunctions);
```

Comment:
I'm surprised we have this information at this point.  Aren't the return funclets actually created during coroutine splitting?

Reply:
The return from an async function (the musttailcall of the parent funclet) is modeled as a separate function pointer argument to the `coro.end.end` async function.

```
   call i1 (ptr, i1, ...) @llvm.coro.end.async(ptr %4, i1 false, ptr @"$s1d3baryySbYaF.0", ptr %21, ptr %22)
   unreachable
  }
  
 define internal swifttailcc void @"$s1d3fooyySbYaF.0"(ptr %0, ptr %1, i1 %2) #8 {
 entry:
   musttail call swifttailcc void %0(ptr swiftasync %1, i1 %2)
   ret void
 }
 ```
 
 We are now keeping these funclet alive instead of inlining it after coro.splitting.

Reply:
Oh.  Is that... what we want to do for this tracing use-case?  So we still have the resume funclet around, and it just immediately tail-calls to this thing that was emitted by the frontend?

Reply:
We can improve upon the generated code in various ways later: e.g. by merging the return funclets into their callers and updating the information during coro splitting, and by not emitting the entries in functions that don't suspend in IRGen.

Reply:
Okay.  And the tracing does record direct tail-calls, despite them just looking like unconditional jumps in assembly?